aspnetcore/performance/response-compression.md

```diff
@@ -0,0 +1,198 @@
+
+## Package
+To include the middleware in your project, add a reference to the `Microsoft.AspNetCore.ResponseCompression` package. The middleware is available for projects that target .NETFramework 4.5.1 or .NETStandard 1.3 or higher.
+
+## Configuration
+### Using defaults
+If you plan to implement the middleware with default GZip compression and for default MIME types (listed below in the [MIME types section](#mime-types)), you can add the middleware to your service collection and processing pipeline. No additional configuration is required. 
+
+[!code-csharp[Main](response-compression/sample/DefaultsSample/Startup.cs?name=snippet1&highlight=3,8)]
+
+Submit a request to the sample application without the `Accept-Encoding` header and observe that the response is uncompressed. The `Content-Encoding` and `Vary` headers are not present on the response.
+
+![Fiddler window showing result of a request without the Accept-Encoding header. The response is not compressed.](response-compression/_static/request-uncompressed.png)
+
+Submit a request to the sample application with the `Accept-Encoding: gzip` header and observe that the response is compressed. The `Content-Encoding` and `Vary` headers are present on the response.
+
+![Fiddler window showing result of a request with the Accept-Encoding header and a value of gzip. The Vary and Content-Encoding headers are added to the response. The response is compressed.](response-compression/_static/request-compressed.png)
+
+### Providers
+#### GzipCompressionProvider
+Use the `GzipCompressionProvider` to compress responses with GZip. This is the default compression provider if none are specified. 
+
+[!code-csharp[Main](response-compression/sample/FullSample/Startup.cs?name=snippet2&highlight=5)]
+
+You can set the compression level with the `GzipCompressionProviderOptions`. The default is for the fastest compression.
+
+Compression Level | Description
+--- | ---
+`CompressionLevel.Fastest` (default) | Compression should complete as quickly as possible, even if the resulting output is not optimally compressed.
+`CompressionLevel.NoCompression` | No compression should be performed.
+`CompressionLevel.Optimal` | Responses should be optimally compressed, even if the compression takes more time to complete.
+
+[!code-csharp[Main](response-compression/sample/FullSample/Startup.cs?name=snippet2&highlight=9-12)]
+
+#### Custom Providers
+You can create custom compression implementations with `ICompressionProvider`. The `encodingName` will reflect the `Accept-Encoding` header value that triggers `CreateStream()` method execution. Using the sample application, the client would submit a request with the `Accept-Encoding: custom` header. The middleware will use the custom compression implementation and return the response with a `Content-Encoding: custom` header. The client must be able to decompress the custom encoding in order for a custom compression implementation to work.
```

Comment:
`CreateStream()` is a pretty specific implementation detail. I would rephrase it as something like "... represent the content-encoding that this `ICompressionProvider` produces. The middleware will use this information to choose the provider based on the list specified in the `Accept-Encoding` header of the request."